lib/kernel.js

```diff
@@ -66,64 +121,124 @@ export default class Kernel {
+          onResults({
+            data: status,
+            stream: "status"
+          });
+        } else {
+          console.warn(
+            "Kernel: ignoring unexpected value for message.content.status"
+          );
+        }
+      } else if (channel === "iopub") {
+        if (message.header.msg_type === "execute_input") {
+          onResults({
+            data: message.content.execution_count,
+            stream: "execution_count"
+          });
+        }
+
+        // TODO(nikita): Consider converting to V5 elsewhere, so that plugins
+        // never have to deal with messages in the V4 format
+        const result = msgSpecToNotebookFormat(msgSpecV4toV5(message));
+        onResults(result);
+      } else if (channel === "stdin") {
+        if (message.header.msg_type !== "input_request") {
+          return;
+        }
+
+        const { prompt } = message.content;
+
+        // TODO(nikita): perhaps it would make sense to install middleware for
+        // sending input replies
+        const inputView = new InputView({ prompt }, (input: string) =>
+          this.kernel.inputReply(input)
+        );
+
+        inputView.attach();
+      }
+    };
   }
 
   complete(code: string, onResults: Function) {
-    throw new Error("Kernel: complete method not implemented");
-  }
-
-  inspect(code: string, curorPos: number, onResults: Function) {
-    throw new Error("Kernel: inspect method not implemented");
-  }
-
-  _parseIOMessage(message: Message) {
-    let result = this._parseExecuteInputIOMessage(message);
-
-    if (!result) {
-      result = msgSpecToNotebookFormat(msgSpecV4toV5(message));
-    }
-
-    return result;
-  }
-
-  _parseExecuteInputIOMessage(message: Message) {
-    if (message.header.msg_type === "execute_input") {
-      return {
-        data: message.content.execution_count,
-        stream: "execution_count"
-      };
-    }
-
-    return null;
+    this.middlewareStack.complete(code, (message: Message, channel: string) => {
+      if (channel !== "shell") {
+        log("Invalid reply: wrong channel");
+        return;
+      }
+      if (!this.kernel._isValidMessage(message)) {
+        //XXX(nikita): it's private!
+        return;
+      }
+      onResults(message.content);
+    });
+  }
+
+  inspect(code: string, cursorPos: number, onResults: Function) {
+    this.middlewareStack.inspect(
+      code,
+      cursorPos,
+      (message: Message, channel: string) => {
+        if (channel !== "shell") {
+          log("Invalid reply: wrong channel");
+          return;
+        }
+        if (!this.kernel._isValidMessage(message)) {
+          //XXX(nikita): it's private!
+          return;
+        }
+        onResults({
+          data: message.content.data,
+          found: message.content.found
+        });
+      }
+    );
   }
 
   destroy() {
-    log("Kernel: Destroying base kernel");
+    log("Kernel: Destroying");
+    this.kernel.destroy();
```

Comment:
Do we want to destroy the kernel before or after removing it from the store?

If before, the subscribers to the store will be notified before the kernel is destroyed and can do something before it's destroyed.